File: TopCoder/src/main/java/com/leetcode/SymmetricTree.java
package com.leetcode;

import com.leetcode.tree.TreeNode;

public class SymmetricTree {


    public boolean isSymmetric(TreeNode root) {
        boolean result = recursive(root.left, root.right);
        return result;
    }

    private boolean recursive(TreeNode root1, TreeNode root2) {
        if (root1 == null && root2 == null) {
            return true;
        }

        if ((root1 == null) || (root2 == null)) {
            return false;
        }

        return root1.val == root2.val
               && recursive(root1.left, root2.right)
               && recursive(root1.right, root2.left);
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(2);
    }
}

File: TopCoder/python/ntree/SymmetricTree.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        result = self.recursive(root.left, root.right)
        return result

    def recursive(self, root1: TreeNode, root2: TreeNode):
        if root1 is None and root2 is None:
            return True

        if root1 is None or root2 is None:
            return False

        return root1.val == root2.val and self.recursive(root1.left, root2.right) and self.recursive(root1.right, root2.left)